src/lib/reporter.test.ts

```diff
@@ -107,41 +145,524 @@ const mockMapDocument: MapDocumentNode = {
   ],
 };
 
+const mockMapDocumentFailure: (provider?: string) => MapDocumentNode = (
+  provider = 'testprovider'
+) => ({
+  kind: 'MapDocument',
+  header: {
+    kind: 'MapHeader',
+    profile: {
+      name: 'test-profile',
+      version: {
+        major: 1,
+        minor: 2,
+        patch: 3,
+      },
+    },
+    provider,
+  },
+  definitions: [
+    {
+      kind: 'MapDefinition',
+      name: 'Test',
+      usecaseName: 'Test',
+      statements: [
+        {
+          kind: 'HttpCallStatement',
+          method: 'GET',
+          url: 'https://unavail.able',
+          request: {
+            kind: 'HttpRequest',
+            security: [],
+          },
+          responseHandlers: [
+            {
+              kind: 'HttpResponseHandler',
+              statusCode: 200,
+              statements: [
+                {
+                  kind: 'OutcomeStatement',
+                  isError: false,
+                  terminateFlow: false,
+                  value: {
+                    kind: 'ObjectLiteral',
+                    fields: [
+                      {
+                        kind: 'Assignment',
+                        key: ['message'],
+                        value: {
+                          kind: 'PrimitiveLiteral',
+                          value: "if you see me, something's wrong",
+                        },
+                      },
+                    ],
+                  },
+                },
+              ],
+            },
+          ],
+        },
+      ],
+    },
+  ],
+});
+
 const mockServer = getLocal();
 
 describe('MetricReporter', () => {
-  it('should report metrics automagically', async () => {
-    SuperJson.loadSync = () => ok(mockSuperJson);
-    jest.mock('../client/registry', () => ({
-      ...jest.requireActual('../client/registry'),
-      getDefaultRegistryUrl: () => mockServer.url,
-    }));
-    await mockServer
-      .post('/sdk-events')
-      .always()
-      .thenCallback(async () => {
-        return {
-          status: 200,
-          body: 'hello',
-        };
-      });
+  let eventEndpoint: MockedEndpoint;
+  beforeEach(async () => {
+    SuperJson.loadSync = () => ok(mockSuperJsonSingle);
+    await mockServer.start();
+    eventEndpoint = await mockServer
+      .post('/insights/sdk_event')
+      .thenJson(202, {});
+    Config().disableReporting = false;
+    Config().superfaceApiUrl = mockServer.url;
+  });
 
+  afterEach(async () => {
+    invalidateSuperfaceClientCache();
+    await mockServer.stop();
+  });
+
+  it('should report SDK Init', async () => {
     const client = new SuperfaceClient();
-    (client as any).boundCache[
-      '{"id":"test-profile","version":"1.0.0"}{"name":"testprovider","security":[]}'
-    ] = new BoundProfileProvider(
+    void client;
+    while (await eventEndpoint.isPending()) {
+      await new Promise(setImmediate);
```

Comment:
How is this different from `Promise.resolve()`?

Reply:
Overall, what's the reason behind these loops?

Reply:
the main difference is that `Promise.resolve()` doesn't work for some reason, it will just loop endlessly until the heat death of the universe. the reason behind the loops is that the reporter calls are void promises, running somewhere in the background with no way to wait for them. if i did just loop, the event loop wouldn't execute, the setImmediate promise await gives it control back for a bit (to resolve any pending calls etc.). that's what i came up with anyway, if you have a better solution tell me

Reply:
Does `Promise.resolve().then(() => undefined)` do something different? The MDN wiki isn't exactly clear. My main issue is that the `setImmediate` api is NodeJS only and it's another hurdle we'd have to get through if we ever get to that sweet browser integration.